components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/HostProxyTest.kt

```diff
@@ -0,0 +1,131 @@
+
+import com.google.common.net.HostAndPort
+import com.hotels.styx.api.Eventual
+import com.hotels.styx.api.HttpRequest
+import com.hotels.styx.api.HttpResponse
+import com.hotels.styx.api.HttpResponseStatus.OK
+import com.hotels.styx.api.LiveHttpRequest
+import com.hotels.styx.client.StyxHostHttpClient
+import com.hotels.styx.routing.RoutingContext
+import com.hotels.styx.routing.handle
+import com.hotels.styx.routing.routingObjectDef
+import io.kotlintest.TestCase
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.FeatureSpec
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import reactor.core.publisher.toMono
+
+class HostProxyTest : FeatureSpec() {
+    val request = HttpRequest.get("/").build()
+    var client: StyxHostHttpClient? = null
+
+    init {
+        feature("Routing and proxying") {
+            scenario("Proxies traffic") {
+                HostProxy(HostAndPort.fromString("localhost:80"), client).let {
+                    it.handle(request.stream(), mockk())
+                }
+
+                verify {
+                    client?.sendRequest(ofType(LiveHttpRequest::class))
+                }
+            }
+
+            scenario("Requests arriving at stopped HostProxy object") {
+                val exception = HostProxy(HostAndPort.fromString("localhost:80"), client).let {
+                    it.stop()
+
+                    shouldThrow<IllegalStateException> {
+                        it.handle(request)
+                                .toMono()
+                                .block()
+                    }
+                }
+
+                exception.message shouldBe ("HostProxy localhost:80 is stopped but received traffic.")
+
+                verify(exactly = 0) {
+                    client?.sendRequest(any())
+                }
+            }
+        }
+
+        feature("HostProxy.Factory") {
+
+            val context = RoutingContext()
+
+            scenario("Uses configured host and port number") {
+                val factory = HostProxy.Factory()
+
+                val hostProxy = factory.build(listOf(), context.get(), routingObjectDef("""
+                          type: HostProxy
+                          config:
+                            host: ahost.server.com:1234
+                        """.trimIndent())) as HostProxy
+
+                hostProxy.hostAndPort.let {
+                    it.hostText shouldBe "ahost.server.com"
+                    it.port shouldBe 1234
+                }
+            }
+
+            scenario("Port defaults to 80") {
+                val factory = HostProxy.Factory()
+
+                val hostProxy = factory.build(listOf(), context.get(), routingObjectDef("""
+                          type: HostProxy
+                          config:
+                            host: localhost
+                        """.trimIndent())) as HostProxy
+
+                hostProxy.hostAndPort.let {
```

Comment:
let() here doesn't seem extremely useful as we are not reading the returned value (e.g, for chaining) or redefining the name of "it" . If we want to reduce verbosity a little bit, we could use run()  for these cases - or apply() if we want to return the receiver object.

Reply:
Fair point. Although I'd prefer `let` because it is also familiar from other languages and therefore bit more self explanatory. Also, without `it` the `run` block might appear bit magical, but also cleaner:

                hostProxy.hostAndPort.run {
                    hostText shouldBe "localhost"
                    port shouldBe 443
                }

I'm happy either way. 

What do you think?

Reply:
I would typically prefer let but in these simple cases run or even apply (I think "run" sounds a bit more obscure )  look neater; however  it adds some complexity in the writing/reading process so it might be easier if we just stick to let{} unless we can't use it.

Reply:
I think it doesn't make a huge difference. I see it's been changed to `run`. I'm fine with that.